### ISSUE TITLE
Reflected operators mistakenly use parent class' non-reflected signature when they should use child class' reflected operator signature

### ISSUE THREAD

**Issue by wwoods** [microsoft/pyright#10644]
**Describe the bug**
Per info box for e.g. `__rrshift__` following, a subclass' reflected operand will be preferentially executed over a parent class' `__rshift__`: https://docs.python.org/3/reference/datamodel.html#object.__rrshift__

Pyright (direct in terminal or in vscode extension) mistakenly uses the parent class' `__rshift__` for type inference.

**Code or Screenshots**

Running the following script with `python` will print R2, indicating a float was returned as expected, but `pyright` will report a type error, that `requires_str` received a string argument.

```python
from typing import Generic, TypeVar

T = TypeVar("T")


class Beep(Generic[T]):
    def __rshift__(self: "Beep[int]", other: "Beep[str]") -> str:
        print("R1")
        return "h"


class Other(Beep[T]):
    def __rrshift__(self: "Other[str]", prev: "Beep[int]") -> float:
        print("R2")
        return 3.


b1 = Beep[int]()
b2 = Other[str]()

b3 = b1 >> b2  
# b3 type inferred as "str", should be "float"
# When ran, R2 is printed as expected

def requires_str(x: float):
    return x + 1.0
requires_str(b3)  # Type error, thinks b3 is str not float
```

**VS Code extension or command-line**
VSCode extension, pyright 1.1.400

**Comment by erictraut** [comment]
Pyright is working as intended here, so I don't consider this a bug.

The runtime behavior described in the documentation that you've linked here unfortunately cannot be replicated reliably during static analysis time. The runtime can determine the precise type of the object and base its "preferential behavior" on that basis. At static analysis time, we don't know the exact type of the operands because static types represent sets of possible values. The type annotation `Beep[int]` includes `Other[int]`, which means that in your example above, the runtime will produce `str` in some cases and `float` in others depending on the actual types at runtime.

I've thought about adding some additional heuristics to pyright's logic here, but they would still be incorrect at times, and they'd add additional complexity and performance overhead in an area that is already complex, so I decided that it wasn't a good tradeoff.

**Comment by erictraut** [comment]
I suppose one could also make the argument that the correct result should be a union of the two results (`str | float`), but most users would find that result even more surprising and likely less useful than what pyright does currently.
